src/main.rs

```diff
@@ -0,0 +1,487 @@
+
+    // make parent directory
+    let mut config_dest = mountpoint.to_path_buf();
+    config_dest.push("ignition");
+    create_dir_all(&config_dest).chain_err(|| "creating Ignition config directory")?;
+
+    // do the copy
+    config_dest.push("config.ign");
+    let mut config_in = OpenOptions::new()
+        .read(true)
+        .open(config_src)
+        .chain_err(|| format!("opening source Ignition config {}", config_src))?;
+    let mut config_out = OpenOptions::new()
+        .write(true)
+        .create_new(true)
+        .open(&config_dest)
+        .chain_err(|| {
+            format!(
+                "opening destination Ignition config {}",
+                config_dest.display()
+            )
+        })?;
+    copy(&mut config_in, &mut config_out).chain_err(|| "writing Ignition config")?;
+
+    Ok(())
+}
+
+/// Write first-boot kernel arguments.
+fn write_firstboot_kargs(mountpoint: &Path, args: &str) -> Result<()> {
+    println!("Writing first-boot kernel arguments");
+
+    // write the arguments
+    let mut config_dest = mountpoint.to_path_buf();
+    config_dest.push("ignition.firstboot");
+    // if the file doesn't already exist, fail, since our assumptions
+    // are wrong
+    let mut config_out = OpenOptions::new()
+        .append(true)
+        .open(&config_dest)
+        .chain_err(|| format!("opening first-boot file {}", config_dest.display()))?;
+    let contents = format!("set ignition_network_kcmdline=\"{}\"\n", args);
+    config_out
+        .write_all(contents.as_bytes())
+        .chain_err(|| "writing first-boot kernel arguments")?;
+
+    Ok(())
+}
+
+/// Override the platform ID.
+fn write_platform(mountpoint: &Path, platform: &str) -> Result<()> {
+    println!("Setting platform to {}", platform);
+
+    // walk /boot/loader/entries/*.conf
+    let mut config_path = mountpoint.to_path_buf();
+    config_path.push("loader/entries");
+    for entry in read_dir(&config_path)
+        .chain_err(|| format!("reading directory {}", config_path.display()))?
+    {
+        let path = entry
+            .chain_err(|| format!("reading directory {}", config_path.display()))?
+            .path();
+        if path.extension().unwrap_or_default() == "conf" {
+            // slurp in the file
+            let mut config = OpenOptions::new()
+                .read(true)
+                .write(true)
+                .open(&path)
+                .chain_err(|| format!("opening bootloader config {}", path.display()))?;
+            let mut orig_contents = String::new();
+            config
+                .read_to_string(&mut orig_contents)
+                .chain_err(|| format!("reading {}", path.display()))?;
+
+            // Rewrite the config.  Assume that we will only install
+            // from metal images and that their bootloader configs will
+            // always set ignition.platform.id.  Fail if those
+            // assumptions change.  This is deliberately simplistic.
```

Comment:
This is fine, though should probably no-op return early on if `platform` is `metal`.